### PR TITLE
Fix DB upgrade

### PR DESCRIPTION
**Pull request by jburel**
# What this PR does

Fix problem when trying to upgrade older DB e.g. nightshade

# Testing this PR

run the upgrade script
The script has already been run to upgrade a copy of nightshade DB to 5.3

# Related reading

cc @joshmoore @mtbc

### CODE REVIEW

File: sql/psql/OMERO5.3__0/OMERO5.2__0.sql [hunk 1]
```diff
@@ -1867,12 +1867,12 @@ ALTER TABLE projectiondef ADD stepping positive_int;
...
 
-ALTER TABLE codomainmapcontext DROP CONSTRAINT FKcodomainmapcontext_renderingDef_renderingdef;
+ALTER TABLE codomainmapcontext DROP CONSTRAINT IF EXISTS FKcodomainmapcontext_renderingDef_renderingdef;
 ALTER TABLE codomainmapcontext DROP renderingdef;
 ALTER TABLE codomainmapcontext DROP renderingdef_index;
 
-DROP TRIGGER codomainmapcontext_renderingDef_index_trigger ON codomainmapcontext;
-DROP FUNCTION codomainmapcontext_renderingDef_index_move();
+DROP TRIGGER IF EXISTS codomainmapcontext_renderingDef_index_trigger ON codomainmapcontext;
+DROP FUNCTION IF EXISTS codomainmapcontext_renderingDef_index_move();
```
Comment (mtbc) [inline sql/psql/OMERO5.3__0/OMERO5.2__0.sql]: in nightshade we do seem to have functions `codomainmapcontext_renderingdef_index_insert` and `codomainmapcontext_renderingdef_index_update`

File: sql/psql/OMERO5.3__0/OMERO5.2__0.sql [hunk 2]
```diff
@@ -1867,12 +1867,12 @@ ALTER TABLE projectiondef ADD stepping positive_int;
...
 -- ... up to patch 10:
 
-ALTER TABLE codomainmapcontext DROP CONSTRAINT FKcodomainmapcontext_renderingDef_renderingdef;
+ALTER TABLE codomainmapcontext DROP CONSTRAINT IF EXISTS FKcodomainmapcontext_renderingDef_renderingdef;
 ALTER TABLE codomainmapcontext DROP renderingdef;
 ALTER TABLE codomainmapcontext DROP renderingdef_index;
 
-DROP TRIGGER codomainmapcontext_renderingDef_index_trigger ON codomainmapcontext;
-DROP FUNCTION codomainmapcontext_renderingDef_index_move();
+DROP TRIGGER IF EXISTS codomainmapcontext_renderingDef_index_trigger ON codomainmapcontext;
```
Comment (mtbc) [inline sql/psql/OMERO5.3__0/OMERO5.2__0.sql:1874]: triggers on nightshade are `codomainmapcontext_renderingdef_index_trigger_insert` and `codomainmapcontext_renderingdef_index_trigger_update`

**Comment by mtbc** [comment]
It will take some time, and some testing on multiple PostgreSQL versions, but tomorrow I could try to figure a way to drop that foreign key constraint by deleting the appropriate row from the `pg_constraint` system catalog.

*Update:* Probably unnecessary, the `ALTER TABLE codomainmapcontext DROP renderingdef` should drop it anyway regardless of its name.

File: sql/psql/OMERO5.3__0/OMERO5.2__0.sql [hunk 3]
```diff
@@ -1867,12 +1867,16 @@ ALTER TABLE projectiondef ADD stepping positive_int;
...
+DROP TRIGGER IF EXISTS codomainmapcontext_renderingDef_index_insert ON codomainmapcontext;
```
Comment (mtbc) [inline sql/psql/OMERO5.3__0/OMERO5.2__0.sql]: needs the `trigger` in the names

**Comment by mtbc** [comment]
This is looking good to me, let's see how testing goes!

**Comment by mtbc** [comment]
Followup to this PR: figure out how these functions/triggers got out of sync and what more maybe needs fixing after 5.3.

**Comment by mtbc** [comment]
After test upgrade from nightshade dump should run `\d codomainmapcontext` in psql and compare with from a newer DB to see if the tables don't have further unsettling discrepancies.

**Comment by jburel** [comment]
@mtbc 
Tested script upgrade from 5.2 DB
```
DROP TRIGGER
NOTICE:  trigger "codomainmapcontext_renderingdef_index_trigger_insert" for relation "codomainmapcontext" does not exist, skipping
DROP TRIGGER
NOTICE:  trigger "codomainmapcontext_renderingdef_index_trigger_update" for relation "codomainmapcontext" does not exist, skipping
DROP TRIGGER
DROP FUNCTION
NOTICE:  function codomainmapcontext_renderingdef_index_insert() does not exist, skipping
DROP FUNCTION
NOTICE:  function codomainmapcontext_renderingdef_index_update() does not exist, skipping

YOU HAVE SUCCESSFULLY UPGRADED YOUR DATABASE TO VERSION OMERO5.3__0
```

**Comment by jburel** [comment]
Tested via breaking too
https://ci.openmicroscopy.org/job/OMERO-DEV-breaking-deploy/851/console
```
INFO  Executing [custom environment]: psql -d OMERO-DEV-breaking-deploy -h localhost -U omero -w -A -t -f /opt/hudson/workspace/OMERO-DEV-breaking-deploy/src/dist/sql/psql/OMERO5.3__0/OMERO5.2__0.sql
2017-03-27 18:10:24,948 [omego.extern] INFO  Completed [54.752 s]
2017-03-27 18:10:24,949 [    omego.db] WARNI stderr: psql:/opt/hudson/workspace/OMERO-DEV-breaking-deploy/src/dist/sql/psql/OMERO5.3__0/OMERO5.2__0.sql:1875: NOTICE:  trigger "codomainmapcontext_renderingdef_index_trigger_insert" for table "codomainmapcontext" does not exist, skipping
psql:/opt/hudson/workspace/OMERO-DEV-breaking-deploy/src/dist/sql/psql/OMERO5.3__0/OMERO5.2__0.sql:1876: NOTICE:  trigger "codomainmapcontext_renderingdef_index_trigger_update" for table "codomainmapcontext" does not exist, skipping
psql:/opt/hudson/workspace/OMERO-DEV-breaking-deploy/src/dist/sql/psql/OMERO5.3__0/OMERO5.2__0.sql:1878: NOTICE:  function codomainmapcontext_renderingdef_index_insert() does not exist, skipping
psql:/opt/hudson/workspace/OMERO-DEV-breaking-deploy/src/dist/sql/psql/OMERO5.3__0/OMERO5.2__0.sql:1879: NOTICE:  function codomainmapcontext_renderingdef_index_update() does not exist, skipping

[TemplateProject] Successfully performed builders from: 'OMERO-DEV-latest-deploy'
```

**Comment by jburel** [comment]
Removed the breaking label and started daily build

**Comment by jburel** [comment]
Tested this morning against nighshade DB
```
NOTICE:  constraint "fkcodomainmapcontext_renderingdef_renderingdef" of relation "codomainmapcontext" does not exist, skipping
ALTER TABLE
ALTER TABLE
ALTER TABLE
NOTICE:  trigger "codomainmapcontext_renderingdef_index_trigger" for relation "codomainmapcontext" does not exist, skipping
DROP TRIGGER
DROP TRIGGER
DROP TRIGGER
NOTICE:  function codomainmapcontext_renderingdef_index_move() does not exist, skipping
```
`` YOU HAVE SUCCESSFULLY UPGRADED YOUR DATABASE TO VERSION OMERO5.3__0``

The script ran smoothly this time

**Comment by mtbc** [comment]
So, from a newer database after upgrade the resulting table looks like,

```
# \d codomainmapcontext
     Table "public.codomainmapcontext"
        Column        |  Type   | Modifiers 
----------------------+---------+-----------
 id                   | bigint  | not null
 permissions          | bigint  | not null
 version              | integer | 
 channelbinding       | bigint  | not null
 creation_id          | bigint  | not null
 external_id          | bigint  | 
 group_id             | bigint  | not null
 owner_id             | bigint  | not null
 update_id            | bigint  | not null
 channelbinding_index | integer | not null
Indexes:
    "codomainmapcontext_pkey" PRIMARY KEY, btree (id)
    "codomainmapcontext_channelbinding_channelbinding_index_key" UNIQUE CONSTRAINT, btree (channelbinding, channelbinding_index)
    "codomainmapcontext_external_id_key" UNIQUE CONSTRAINT, btree (external_id)
    "i_codomainmapcontext_channelbinding" btree (channelbinding)
    "i_codomainmapcontext_group" btree (group_id)
    "i_codomainmapcontext_owner" btree (owner_id)
Foreign-key constraints:
    "fkcodomainmapcontext_channelbinding_channelbinding" FOREIGN KEY (channelbinding) REFERENCES channelbinding(id)
    "fkcodomainmapcontext_creation_id_event" FOREIGN KEY (creation_id) REFERENCES event(id)
    "fkcodomainmapcontext_external_id_externalinfo" FOREIGN KEY (external_id) REFERENCES externalinfo(id)
    "fkcodomainmapcontext_group_id_experimentergroup" FOREIGN KEY (group_id) REFERENCES experimentergroup(id)
    "fkcodomainmapcontext_owner_id_experimenter" FOREIGN KEY (owner_id) REFERENCES experimenter(id)
    "fkcodomainmapcontext_update_id_event" FOREIGN KEY (update_id) REFERENCES event(id)
Referenced by:
    TABLE "contraststretchingcontext" CONSTRAINT "fkcontraststretchingcontext_codomainmapcontext_id_codomainmapco" FOREIGN KEY (codomainmapcontext_id) REFERENCES codomainmapcontext(id)
    TABLE "planeslicingcontext" CONSTRAINT "fkplaneslicingcontext_codomainmapcontext_id_codomainmapcontext" FOREIGN KEY (codomainmapcontext_id) REFERENCES codomainmapcontext(id)
    TABLE "reverseintensitycontext" CONSTRAINT "fkreverseintensitycontext_codomainmapcontext_id_codomainmapcont" FOREIGN KEY (codomainmapcontext_id) REFERENCES codomainmapcontext(id)
Triggers:
    codomainmapcontext_channelbinding_index_trigger BEFORE UPDATE ON codomainmapcontext FOR EACH ROW EXECUTE PROCEDURE codomainmapcontext_channelbinding_index_move()
```

Barring perhaps some weirdly named foreign key constraints does this sufficiently match that from the upgraded nightshade database?

**Comment by jburel** [comment]
After nightshade upgrade
```
\d codomainmapcontext
     Table "public.codomainmapcontext"
        Column        |  Type   | Modifiers 
----------------------+---------+-----------
 id                   | bigint  | not null
 owner_id             | bigint  | not null
 group_id             | bigint  | not null
 creation_id          | bigint  | not null
 update_id            | bigint  | not null
 permissions          | bigint  | not null
 external_id          | bigint  | 
 version              | integer | default 0
 channelbinding       | bigint  | not null
 channelbinding_index | integer | not null
Indexes:
    "codomainmapcontext_pkey" PRIMARY KEY, btree (id)
    "codomainmapcontext_channelbinding_channelbinding_index_key" UNIQUE CONSTRAINT, btree (channelbinding, channelbinding_index)
    "codomainmapcontext_external_id_key" UNIQUE CONSTRAINT, btree (external_id)
    "i_codomainmapcontext_channelbinding" btree (channelbinding)
    "i_codomainmapcontext_group" btree (group_id)
    "i_codomainmapcontext_owner" btree (owner_id)
Foreign-key constraints:
    "fk9ad375033680cece" FOREIGN KEY (owner_id) REFERENCES experimenter(id)
    "fk9ad375033ebb3c78" FOREIGN KEY (creation_id) REFERENCES event(id)
    "fk9ad3750354212865" FOREIGN KEY (external_id) REFERENCES externalinfo(id)
    "fk9ad3750397e49cf3" FOREIGN KEY (group_id) REFERENCES experimentergroup(id)
    "fk9ad37503e046324e" FOREIGN KEY (update_id) REFERENCES event(id)
    "fkcodomainmapcontext_channelbinding_channelbinding" FOREIGN KEY (channelbinding) REFERENCES channelbinding(id)
Referenced by:
    TABLE "contraststretchingcontext" CONSTRAINT "fk17c6119027e86052" FOREIGN KEY (codomainmapcontext_id) REFERENCES codomainmapcontext(id)
    TABLE "reverseintensitycontext" CONSTRAINT "fk7c455cfe27e86052" FOREIGN KEY (codomainmapcontext_id) REFERENCES codomainmapcontext(id)
    TABLE "planeslicingcontext" CONSTRAINT "fk96bf2a5c27e86052" FOREIGN KEY (codomainmapcontext_id) REFERENCES codomainmapcontext(id)
Triggers:
    codomainmapcontext_channelbinding_index_trigger BEFORE UPDATE ON codomainmapcontext FOR EACH ROW EXECUTE PROCEDURE codomainmapcontext_channelbinding_index_move()

```

**Comment by jburel** [comment]
Also saved some rendering settings from insight.
```
select * from codomainmapcontext;
 id | owner_id | group_id | creation_id | update_id | permissions | external_id | version | channelbinding | channelbinding_index 
----+----------+----------+-------------+-----------+-------------+-------------+---------+----------------+----------------------
  1 |       16 |        3 |    54285464 |  54285464 |         -56 |             |       0 |        2998028 |                    0
```

**Pull request merged**